src/main/kotlin/app/hashers/CommitHasher.kt

```diff
@@ -3,240 +3,99 @@
             Logger.debug("Sent ${commits.size} added commits to server")
         }
     }
 
-    private fun postFactsToServer(facts: List<Fact>) {
-        if (facts.isNotEmpty()) {
-            api.postFacts(facts)
-            Logger.debug("Sent ${facts.size} facts to server")
-        }
-    }
-
     private fun deleteCommitsOnServer(commits: List<Commit>) {
         if (commits.isNotEmpty()) {
             api.deleteCommits(commits)
             Logger.debug("Sent ${commits.size} deleted commits to server")
         }
     }
 
-    private fun getCommitsAsObservable(): Observable<Commit> =
-        Observable.create { subscriber ->
-        try {
-            val revWalk = RevWalk(gitRepo)
-            val commitId = gitRepo.resolve(RepoHelper.MASTER_BRANCH)
-            revWalk.markStart(revWalk.parseCommit(commitId))
-            for (revCommit in revWalk) {
-                subscriber.onNext(Commit(revCommit))
-            }
-        } catch (e: Exception) {
-            Logger.error("Commit producing error", e)
-            subscriber.onError(e)
-        }
-        subscriber.onComplete()
-    }
-
     private val emailFilter: (Commit) -> Boolean = {
         val email = it.author.email
         localRepo.hashAllContributors || (email == localRepo.author.email ||
-            repo.emails.contains(email))
+            serverRepo.emails.contains(email))
     }
 
-    private fun <T> Observable<T>.pairWithNext(): Observable<Pair<T, T>> {
-        return this.map { emit -> Pair(emit, emit) }
-                    // Accumulate emits by prev-next pair.
-                    .scan { pairAccumulated, pairNext ->
-                        Pair(pairAccumulated.second, pairNext.second)
-                    }
-                    .skip(1)  // Skip initial not paired emit.
-    }
+    // Hash added and missing server commits and send them to server.
+    fun updateFromObservable(observable: Observable<Commit>) {
+        val lastKnownCommit = serverRepo.commits.lastOrNull()
+        val knownCommits = serverRepo.commits.toHashSet()
 
-    fun update() {
-        // Delete locally missing commits from server. If found at least one
-        // common commit then next commits are not deleted because hash of a
-        // commit calculated including hashes of its parents.
-        val firstOverlapCommit = findFirstOverlappingCommit()
-        val deletedCommits = repo.commits
-            .takeWhile { it.rehash != firstOverlapCommit?.rehash }
-        deleteCommitsOnServer(deletedCommits)
+        val throwables = mutableListOf<Throwable>()
 
-        // Hash added and missing server commits and send them to server.
-        hashAndSendCommits()
+        observable
+            .takeWhile { new ->  // Hash until last known commit.
+                new.rehash != lastKnownCommit?.rehash
+            }
+            .filter { commit ->  // Don't hash known.
+                knownCommits.isEmpty() || !knownCommits.contains(commit)
```

Comment:
how could we reach it if we do takeWhile() until the lastKnownCommit?

Reply:
Server could miss some of the commits.